### PR TITLE
Refactoring hoomd polymers 

### PR DESCRIPTION
**Pull request by marjanAlbouye**
This PR refactors the base classes in the repository with the goal of making the whole framework flexible and reusable.

Currently, the bulk of refactoring is going to happen on Molecule and System classes.

### CODE REVIEW

File: hoomd_polymers/base/molecule.py [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+            return mb.load(self.smiles, smiles=True)
+        else:
+            raise ValueError(
+                    "Unable to load from ",
+                    f"File: {self.file}",
+                    f"SMILES: {self.smiles}"
+            )
+
+    def _generate(self):
+        self._molecules.append(self._mb_molecule)
```
Comment (chrisjonesBSU) [inline hoomd_polymers/base/molecule.py:85]: Do we need a for loop here using `n_mols`? 

```
for i in range(self.n_mols):
    self._molecules.append(self._mb_molecule)
```

Comment (marjanAlbouye) [reply hoomd_polymers/base/molecule.py:85]: Yeah, good catch! I'll fix it.

File: hoomd_polymers/base/molecule.py [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+                    f"File: {self.file}",
+                    f"SMILES: {self.smiles}"
+            )
+
+    def _generate(self):
+        for i in range(self.n_mols):
+            self._molecules.append(self._mb_molecule)
+
+    def _convert_to_gmso(self):
+        topology = from_mbuild(self._mb_molecule)
```
Comment (chrisjonesBSU) [inline hoomd_polymers/base/molecule.py:97]: We need to be careful here. Related to what we've talked about befre with some of the espaloma challenges is how new types of particles, bonds and angles might emerge once we polymerize a monomer.  For example, the PPS monomer doesn't have a C-S-C angle, but the polymer does.  Note, the monomer has a C-C-S angle, but that is different than a C-S-C angle.

So, it's possible that we aren't completely validating a FF if we just use the monomer, and the `topology_information' might return incomplete information.  Maybe we just pull the first compound from `self.molecules` and try using that?

File: hoomd_polymers/base/molecule.py [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+            bond_types.add(tuple(bond_connections))
+        return bond_types
+
+    def _identify_angle_types(self, gmso_molecule):
+        angle_types = set()
+        for angle in gmso_molecule.angles:
+            angle_connections = [angle.connection_members[0].atom_type or angle.connection_members[0].name,
+                                 angle.connection_members[1].atom_type or angle.connection_members[1].name,
+                                 angle.connection_members[2].atom_type or angle.connection_members[2].name]
+            angle_connections.sort()
```
Comment (chrisjonesBSU) [inline hoomd_polymers/base/molecule.py:129]: The order of the angles is important, so I don't think we should use sort in this way.  For example, a C-S-C angle is different than a C-C-S angle, and a molecule might have both of them.

File: hoomd_polymers/base/molecule.py [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+        return angle_types
+
+    def _identify_dihedral_types(self, gmso_molecule):
+        dihedral_types = set()
+        for dihedral in gmso_molecule.dihedrals:
+            dihedral_connections = [dihedral.connection_members[0].atom_type or dihedral.connection_members[0].name,
+                                    dihedral.connection_members[1].atom_type or dihedral.connection_members[1].name,
+                                    dihedral.connection_members[2].atom_type or dihedral.connection_members[2].name,
+                                    dihedral.connection_members[3].atom_type or dihedral.connection_members[3].name]
+            dihedral_connections.sort()
```
Comment (chrisjonesBSU) [inline hoomd_polymers/base/molecule.py:140]: See comment above about sorting angles; same thing applies for dihedrals.

File: hoomd_polymers/base/molecule.py [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+        return dihedral_types
+
+    def _identify_improper_types(self, gmso_molecule):
+        improper_types = set()
+        for improper in gmso_molecule.impropers:
+            improper_connections = [improper.connection_members[0].atom_type or improper.connection_members[0].name,
+                                    improper.connection_members[1].atom_type or improper.connection_members[1].name,
+                                    improper.connection_members[2].atom_type or improper.connection_members[2].name,
+                                    improper.connection_members[3].atom_type or improper.connection_members[3].name]
+            improper_connections.sort()
```
Comment (chrisjonesBSU) [inline hoomd_polymers/base/molecule.py:151]: Same thing as the previous 2 comments.

File: hoomd_polymers/base/molecule.py (hunk 2 above)
Comment (marjanAlbouye) [reply hoomd_polymers/base/molecule.py:97]: Yeah, that's a good point. I'll use the first compound from molecules list instead.

File: hoomd_polymers/base/molecule.py (hunk 3 above)
Comment (chrisjonesBSU) [reply hoomd_polymers/base/molecule.py:129]: @marjanAlbouye I don't think we need to worry too much about doing any sorting here. It's important that everything is sorted correctly when creating the hoomd force objects, but gmso does that in the `convert_hoomd` stuff.  Just having non-sorted types of angles and dihedrals present is probably good enough for what we're trying to do here in just providing a quick summary of the types of BADs.

**Comment by chrisjonesBSU** [comment]
I don't think the topology information attributes are returning sets like we expect them to.


For example:

```
from hoomd_polymers.library.polymers import PPS

>>>pps = PPS(n_mols=5, lengths=20, force_field="pps_opls")
>>>len(pps.bond_types)
241
```

I don't think the `_identify_bond_types()` function is creating a set as expected, because the information has the object IDs in it, so no two bonds are the same.

for example:

```
>>> print(pps.bond_types)

{(<AtomType ca,
   expression: 4*epsilon*(-sigma**6/r**6 + sigma**12/r**12),
   id: 6632899968,
   atomclass: CA>,
  <AtomType ca,
   expression: 4*epsilon*(-sigma**6/r**6 + sigma**12/r**12),
   id: 6632900736,
   atomclass: CA>),
...
}
```

**Comment by chrisjonesBSU** [comment]
Ok, populating sets of the topology info should be fixed.  `atom_type` in GMSO shows the atom type name, the expression used for the pair potential and the memory ID of the atom object.  Changing `atom_type` to `atom_type.name` fixes this.

**Comment by chrisjonesBSU** [comment]
I think this is getting close. We just need unit tests.  

@marjanAlbouye Can you think of anything else we are missing that we wanted in the molecule class?  I think the `CoPolymer()` class needs some attention, I can take a look at that.

If not, I think the next step is to merge this, then start a branch to finish refactoring the system classes?

**Comment by chrisjonesBSU** [comment]
> I still need to implement hoomd forcefield validation function. I will do that this week. I also need to double check System class and see if there is anything needed there. But other than that I think this PR is ready to be merged.

I think there is still quite a bit to do in `system.py`. We need to figure out how to handle and apply forcefields with molecules now holding FF information. We can keep this open and work on the System stuff

**Pull request merged**